### PR TITLE
fix memory leak in swt_axis

### PR DESCRIPTION
**Pull request by grlee77**
Previous to this patch, `output_dims.shape` and `output_dims.strides` could end up pointing to an invalid memory location.  I have moved the setting of these attributes inside the levels loop so that they are always pointing to a current array.  

This allows removing one initial `np.empty` call prior to that loop.  I also change the `cA` and `cD` initializations to use `np.empty` rather than `np.zeros` for efficiency.

For the SWT, `input_shape` and `output_shape` are always the same, so I removed the `input_shape` variable.  I initially tried removing `output_dims` and just using `data_dims` in its place, but this breaks any case where the input array is not C-contiguous as the strides are then different.  (The outputs of this function are C-contiguous, but the input does not have to be).

closes gh-288

### CODE REVIEW

**Comment by grlee77** [comment]
@rgommers 
Can you think of a good way to test this?  I verified that I no longer get the `RuntimeError` for the example in #288 after this fix.  It is hard to put that in a test, though, because occasionally it does happen to work without error even for the old code.  (Also, we would have to add a dependency on the `futures` package to be able to use `concurrent.futures` in Python 2.7)

**Comment by rgommers** [comment]
> because occasionally it does happen to work without error even for the old code. 

A 100% reliable test for this kind of thing isn't possible, so I'd just go with that example from gh-288. If it fails most of the time, that's good enough if the test is run often (there's 14 test suite runs per PR).

If you skip the test if ``futures`` is not installed, and add a ``pip install futures`` to the TravisCI and Appveyor script, then the dependency isn't much of an issue.

**Comment by grlee77** [comment]
sounds reasonable.  will update when I get a chance.  It should probably be sufficient to test a loop like in gh-288, for `swtn`, 'wavedecn' for 2D data and `swt`, `wavedec` for 1D data.  That would test the 4 different corresponding cython routines: `swt_axis`, `dwt_axis`, `swt`, `dwt_single`.  The 2D cases just end up calling the nD functions so probably don't need separate tests.

**Comment by grlee77** [comment]
See how this looks.  I ended up deciding to test all `dwt*`, `swt*`, `wavedec*` as well as `cwt`.  The range for the loops was adjusted to keep the time down to approximately 1 second for each of the new tests.

File: pywt/_extensions/_swt.pyx [hunk 1]
```diff
@@ -184,15 +178,18 @@ cpdef swt_axis(np.ndarray data, Wavelet wavelet, size_t level,
...
             raise TypeError("Array must be floating point, not {}"
                             .format(data.dtype))
         ret.append((cA, cD))
 
         # previous approx coeffs are the data for the next level
         data = cA
-        data_info = output_info
+        # update data_info to match the new data array
+        data_info.strides = <pywt_index_t *> data.strides
+        data_info.shape = <size_t *> data.shape
```
Comment (rgommers) [inline pywt/_extensions/_swt.pyx]: ``data_info`` is overwritten at the top of this for-loop, so these lines do nothing.

Comment (rgommers) [inline pywt/_extensions/_swt.pyx]: eh never mind, it's late and I can't read - deleting comment again

**Comment by rgommers** [comment]
Test looks quite good, very comprehensive coverage. Merging, thanks @grlee77

**Pull request merged**